packages/compiler-cli/test/ngtsc/ngtsc_spec.ts

```diff
@@ -3243,6 +3243,214 @@ runInEachFileSystem(os => {
+        env.write('dir.ts', `
+          import {Directive} from '@angular/core';
+
+          @Directive({
+            selector: 'dir',
+          })
+          export class Dir {}
+        `);
+        env.write('module.ts', `
+          import {NgModule} from '@angular/core';
+          import {Dir} from './dir';
+
+          @NgModule({
+            declarations: [Dir],
+            exports: [Dir],
+          })
+          export class Module {}
+
+          export {Dir};
+        `);
+
+        env.driveMain();
+        const jsContents = env.getContents('module.js');
+        const dtsContents = env.getContents('module.d.ts');
+
+        expect(jsContents).not.toContain('export { Dir as ɵngExportɵModuleɵDir } from "./dir";');
+        expect(dtsContents).not.toContain('export { Dir as ɵngExportɵModuleɵDir } from "./dir";');
+      });
+
+      it('should error when two directives with the same declared name are exported from the same NgModule',
+         () => {
+           env.write('dir.ts', `
+          import {Directive} from '@angular/core';
+
+          @Directive({
+            selector: 'dir',
+          })
+          export class Dir {}
+        `);
+           env.write('dir2.ts', `
+          import {Directive} from '@angular/core';
+
+          @Directive({
+            selector: 'dir',
+          })
+          export class Dir {}
+        `);
+           env.write('module.ts', `
+          import {NgModule} from '@angular/core';
+          import {Dir} from './dir';
+          import {Dir as Dir2} from './dir2';
+
+          @NgModule({
+            declarations: [Dir, Dir2],
+            exports: [Dir, Dir2],
+          })
+          export class Module {}
+        `);
+
+           const diag = env.driveDiagnostics();
+           expect(diag.length).toBe(1);
+           expect(diag[0] !.code)
+               .toEqual(Number.parseInt(`-99${ErrorCode.NGMODULE_REEXPORT_NAME_COLLISION}`));
+         });
+
+      it('should not error when two directives with the same declared name are exported from the same NgModule, but one is exported from the file directly',
+         () => {
+           env.write('dir.ts', `
+             import {Directive} from '@angular/core';
+   
+             @Directive({
+               selector: 'dir',
+             })
+             export class Dir {}
+           `);
+           env.write('dir2.ts', `
+             import {Directive} from '@angular/core';
+   
+             @Directive({
+               selector: 'dir',
+             })
+             export class Dir {}
+           `);
+           env.write('module.ts', `
+             import {NgModule} from '@angular/core';
+             import {Dir} from './dir';
+             import {Dir as Dir2} from './dir2';
+   
+             @NgModule({
+               declarations: [Dir, Dir2],
+               exports: [Dir, Dir2],
+             })
+             export class Module {}
+
+             export {Dir} from './dir2';
```

Comment:
This testcase would be a bit more interesting if this were
```suggestion
             export {Dir2};
```

Reply:
Personally, I find it more interesting the way it is :grin: